cocotb/share/lib/simulator/simulatormodule.cpp

```diff
@@ -41,11 +41,127 @@ static int sim_ending = 0;
+    }
+    gpi_hdl_Object<gpi_hdl> *obj = PyObject_New(gpi_hdl_Object<gpi_hdl>, &gpi_hdl_Object<gpi_hdl>::py_type);
+    if (obj == NULL) {
+        return NULL;
+    }
+    obj->hdl = hdl;
+    return (PyObject *)obj;
+}
+
+template<typename gpi_hdl>
+static PyObject *gpi_hdl_richcompare(PyObject *self, PyObject *other, int op) {
+    gpi_hdl_Object<gpi_hdl> *self_hdl_obj;
+    gpi_hdl_Object<gpi_hdl> *other_hdl_obj;
+
+    if (
+            Py_TYPE(self) != &gpi_hdl_Object<gpi_hdl>::py_type ||
+            Py_TYPE(other) != &gpi_hdl_Object<gpi_hdl>::py_type) {
+        Py_RETURN_NOTIMPLEMENTED;
+    }
+
+    self_hdl_obj = (gpi_hdl_Object<gpi_hdl> *)self;
+    other_hdl_obj = (gpi_hdl_Object<gpi_hdl> *)other;
+
+    switch (op) {
+        case Py_EQ: return PyBool_FromLong(self_hdl_obj->hdl == other_hdl_obj->hdl);
+        case Py_NE: return PyBool_FromLong(self_hdl_obj->hdl != other_hdl_obj->hdl);
+        default: Py_RETURN_NOTIMPLEMENTED;
+    }
+}
+
+// Slots which vary ebtween the extension types
+template<typename gpi_hdl>
+struct gpi_hdl_type_slots {
+    static char const* const name;
+};
+
+template<> char const* const gpi_hdl_type_slots<gpi_sim_hdl>::name = "gpi_sim_hdl";
+template<> char const* const gpi_hdl_type_slots<gpi_cb_hdl>::name = "gpi_cb_hdl";
+template<> char const* const gpi_hdl_type_slots<gpi_iterator_hdl>::name = "gpi_iterator_hdl";
+
+// Initialize the python type slots
+template<typename gpi_hdl> PyTypeObject gpi_hdl_Object<gpi_hdl>::py_type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    /* tp_name              */ gpi_hdl_type_slots<gpi_hdl>::name,
+    /* tp_basicsize         */ sizeof(gpi_hdl_Object<gpi_hdl>),
+    /* tp_itemsize          */ 0,
+    /* tp_dealloc           */ 0,
+    /* tp_vectorcall_offset */ 0,
+    /* tp_getattr           */ 0,
+    /* tp_setattr           */ 0,
+    /* tp_as_async          */ 0,
+    /* tp_repr              */ (reprfunc)gpi_hdl_repr<gpi_hdl>,
+    /* tp_as_number         */ 0,
+    /* tp_as_sequence       */ 0,
+    /* tp_as_mapping        */ 0,
+    /* tp_hash              */ (hashfunc)gpi_hdl_hash<gpi_hdl>,
+    /* tp_call              */ 0,
+    /* tp_str               */ 0,
+    /* tp_getattro          */ 0,
+    /* tp_setattro          */ 0,
+    /* tp_as_buffer         */ 0,
+    /* tp_flags             */ Py_TPFLAGS_DEFAULT,
+    /* tp_doc               */ 0,
+    /* tp_traverse          */ 0,
+    /* tp_clear             */ 0,
+    /* tp_richcompare       */ gpi_hdl_richcompare<gpi_hdl>,
+    /* tp_weaklistoffset    */ 0,
+    /* tp_iter              */ 0,
+    /* tp_iternext          */ 0,
+    /* tp_methods           */ 0,
+    /* tp_members           */ 0,
+    /* tp_getset            */ 0,
+    /* tp_base              */ 0,
+    /* tp_dict              */ 0,
+    /* tp_descr_get         */ 0,
+    /* tp_descr_set         */ 0,
+    /* tp_dictoffset        */ 0,
+    /* tp_init              */ 0,
+    /* tp_alloc             */ 0,
+    /* tp_new               */ 0,
+};
```

Comment:
You might be interested in [this](https://github.com/googleprojectzero/functionsimsearch/blob/ec5d9e1224ff915b3dc9e7af19e21e110c25239c/pybindings/pybindings.cpp#L346-L358) as a way around this.

Reply:
That's a nice trick, thanks

Reply:
Adopted. Needed slight modification not to fail due to uninitialized field warnings.